### ISSUE TITLE
Getting MD5 checksum through Digest::MD5 might fail

### ISSUE THREAD

**Issue by ferki** [RexOps/Rex#1440]
## Describe the bug

According to [CPAN Testers matrix for Rex-1.13.0](http://matrix.cpantesters.org/?dist=Rex%201.13.0), getting MD5 checksum of files through Digest::MD5 might fail under some circumstances. The 3 failing reports so far:

- http://www.cpantesters.org/cpan/report/0eaa4966-0744-11eb-9d09-01681f24ea8f
- http://www.cpantesters.org/cpan/report/0a7a8d92-0744-11eb-9d09-01681f24ea8f
- http://www.cpantesters.org/cpan/report/06551796-0744-11eb-9d09-01681f24ea8f

What seems to be common in these cases at first glance is they use Digest::MD5-2.56 (released yesterday), and they run on Debian Wheezy (which is not supported by its upstream anymore).

## How to reproduce it

It's unclear how exactly to reproduce it yet, using Digest::MD5-2.56 or later seems to work fine on my systems and in CI.

1. Setup the environment like in the failing reports (not sure what's the crucial factor though)
2. Run the MD5 tests: `prove -lrv t/md5.t`

~~Shortest code example that demonstrates the bug:~~

## Expected behavior

Digest::MD5 is a dependency of Rex, so it is expected to be provide proper results during local test runs.

## Circumstances

 - Rex version: 1.13.0
 - Perl version: multiple (5.20.0, 5.22.2, 5.24.0)
 - OS running rex: Debian Wheezy
 - OS managed by rex: local
 - How rex was installed: cpan

~~## Debug log~~

**Comment by ferki** [comment]
@eserte: The failing CPAN Testers reports about this issue seems to come only from one of your machines (thanks! :heart:). Could you perhaps help with some additional insights to reproduce it, please?

I tried various combinations of the same older perls with Digest::MD5 versions with perlbrew, but no luck so far. Maybe it's more related to something specific to Debian Wheezy?

**Comment by ferki** [comment]
Alright, there are now matching PASS reports (thank you!):

- http://www.cpantesters.org/cpan/report/9dda2678-0826-11eb-aff3-ca161f24ea8f
- http://www.cpantesters.org/cpan/report/55ec548a-0826-11eb-84c9-19891f24ea8f
- http://www.cpantesters.org/cpan/report/32719114-0826-11eb-a123-967f1f24ea8f

All the FAIL reports seemed to have had this environment variable set:

```
PERL5LIB="/home/cpansand/.cpan/build/2020100519/Digest-MD5-2.56-0/blib/arch:/home/cpansand/.cpan/build/2020100519/Digest-MD5-2.56-0/blib/lib"
```

while the PASSing ones did not.

Since I can't reproduce it, and there are PASSing tests from the same machine, I'm closing this for now.

**Comment by eserte** [comment]
@ferki: Well, things are actually not well... it seems that t/md5.t everywhere fails if newest Digest::MD5 (2.58) is installed.

**Comment by eserte** [comment]
No, it must be something different. There are also pass reports with Digest::MD5 2.58 active.

Next theory: it fails if Digest::MD5 is *uninstalled*, that is, the module is used from the blib variant in PERL5LIB. I think it can be reproduced by doing the following in a CPAN.pm shell (untested):
```
test Digest::MD5
test Rex
```

**Comment by ferki** [comment]
@eserte: thanks for the feedback. I have Digest::MD5-2.58 locally and it passes; apparently for others on CPAN Testers too. I'm interested in any failure reports, if you have one later!

As far as unavailable Digest::MD5 is concerned, that function might return `undef` correctly in this case. But Digest::MD5 is a dependency of Rex, so I'd expect it is available under standard CPAN Testers circumstances. How it could go missing between installing all the dependencies and running the tests?

**Comment by eserte** [comment]
@ferki: the problem is that `t/md5.t` uses a "wrong" perl somewhere. I can reproduce the problem by using the CPAN.pm shell with a perl which is not the system perl (especially not the first perl in user's PATH) and which is "incompatible" in some way (either it is newer, or is compiled differently with regards to -Dusethreads), using the two CPAN shell lines already mentioned above.

Running the test with strace shows the problem:
```
29568 08:52:06.648270 write(2, "Can't load '/home/e/eserte/.cpan/build/Digest-MD5-2.58-0/blib/arch/auto/Diges/MD5/MD5.so' for module Digest::MD5: /home/e/eserte/.cpan/build/Digest-MD5-2.58-0/blib/arch/auto/Digest/MD5/MD5.so: undefined symbol: PL_stack_sp at /usr/lib/x86_64-linux-gnu/perl/5.20/DynaLoader.pm line 187.\n at -e line 0.\nCompilation failed in require.\nBEGIN failed--compilation aborted.\n", 368) = 368 <0.000011>
```

The usual fix is to not call "perl", but `$^X` in these situations.

**Comment by ferki** [comment]
> @ferki: the problem is that `t/md5.t` uses a "wrong" perl somewhere. I can reproduce the problem by using the CPAN.pm shell with a perl which is not the system perl (especially not the first perl in user's PATH) and which is "incompatible" in some way (either it is newer, or is compiled differently with regards to -Dusethreads), using the two CPAN shell lines already mentioned above.

Thanks for the explanation. I'm glad the new tests are catching such corner cases that were hidden previously. I'm still trying to understand if this is a common and/or normal environment or not, though :)

> The usual fix is to not call "perl", but `$^X` in these situations.

Hmm, yes, that might work better for local calls. That is to make sure the same perl is used in any potential command runs, where Digest::MD5 is already a dependency through Rex itself.

**Comment by ferki** [comment]
@eserte: I've seen you've generated _lots of_ FAIL reports by now :upside_down_face:

Could you try testing a hotfix [patch](a311a7fb5fb19610bf528e4bf08af6861ac4d9da) from the `md5` branch too, please? If that works, we can either merge that already, or even make the idea into a generic wrapper to be used throughout the codebase.

**Comment by eserte** [comment]
Seems to work: run CPAN shell, test Digest::MD5, open a shell from CPAN.pm, run "dzil test" on master → fails, run the same on branch md5 → passes. Though a dev/trial release on CPAN would be more conclusive.

**Comment by ferki** [comment]
Thanks again @eserte for the testing and feedback! I've now released 1.13.0.1-TRIAL to CPAN with the proposed fix to gather even more results.
